File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.ans = 0
    def bt(self, node, depth):
        l, r = 0, 0
        if node.left != None:
            l = self.bt(node.left, depth + 1)
        if node.right != None:
            r = self.bt(node.right, depth + 1)
        if node.left == None and node.right == None: 
            return depth
        self.ans = max(self.ans, (l - depth) + (r - depth) )
        return max(l, r)
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root == None: return 0
        _ = self.bt(root, 0)
        return self.ans

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;

        for (int i = 0; i < heights.size(); i++) {
            int left = i;
            while (left >= 0 && heights[left] >= heights[i]) {
                left--;
            }
            int right = i;
            while (right < heights.size() && heights[right] >= heights[i]) {
                right++;
            }

            left++;
            right--;
            int area = (right - left + 1) * heights[i];
            res = max(res, area);
        }
        return res;
    }
};
